physical/raft/fsm.go

```diff
@@ -633,6 +637,116 @@ func listPageInner(ctx context.Context, tx *bolt.Tx, prefix string, after string
 	return keys, nil
 }
 
+func (f *FSM) applyBatchNonTxOps(b *bolt.Bucket, txnState *fsmTxnCommitIndexApplicationState, command *LogData) error {
+	for _, op := range command.Operations {
+		var err error
+		switch op.OpType {
+		case putOp:
+			err = b.Put([]byte(op.Key), op.Value)
+			if err == nil {
+				txnState.logWrite(op.Key)
+			}
+		case deleteOp:
+			err = b.Delete([]byte(op.Key))
+			if err == nil {
+				txnState.logWrite(op.Key)
+			}
+		case restoreCallbackOp:
+			if f.restoreCb != nil {
+				// Kick off the restore callback function in a go routine
+				go f.restoreCb(context.Background())
+			}
+		default:
+			if _, ok := f.unknownOpTypes.Load(op.OpType); !ok {
+				f.logger.Error("unsupported transaction operation", "op", op.OpType)
+				f.unknownOpTypes.Store(op.OpType, struct{}{})
+			}
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *FSM) applyBatchTxOps(tx *bolt.Tx, b *bolt.Bucket, txnState *fsmTxnCommitIndexApplicationState, command *LogData) error {
+	txnState.setInTx()
+
+	// First we verify the transaction can apply correctly before doing any
```

Comment:
@JanMa I see, in order to reorder this, we have to reorder them in the original verification log as well, so they operate off of the pre-commit data, not the post-commit data... This will take a bit more thinking to do.